### PR TITLE
[IMP] mail: enable inheritance in the next activity values computation

### PR DESCRIPTION
**Pull request by h4818**
values for next activities are currently prepared inside the `mail_activity._action_done()` method which makes inheritance very difficult. This commit delegates the value computation to a sub-method and thereby improves the extensibility of the module.  

required to support changes to enterprise documents odoo/enterprise#20769 
task-2627837

### CODE REVIEW

File: addons/mail/models/mail_activity.py [hunk 1]
```diff
@@ -504,6 +504,25 @@ def action_feedback_schedule_next(self, feedback=False):
             'target': 'new',
         }
 
+    def _prepare_next_activity_values(self):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_activity.py:507]: Organization: would move it at the end of file, under a 
```
# ----------------------------------------------------------------------
# TOOLS
# ----------------------------------------------------------------------
```
section so that it does not mess with current list of main action methods. Purpose is to keep main API on top.

File: addons/mail/models/mail_activity.py [hunk 2]
```diff
@@ -504,6 +504,25 @@ def action_feedback_schedule_next(self, feedback=False):
...
+        """
+        self.ensure_one()
+        vals = self.default_get(self.fields_get())
+
+        vals.update({
+            'previous_activity_type_id': self.activity_type_id.id,
+            'res_id': self.res_id,
+            'res_model': self.res_model,
+            'res_model_id': self.env['ir.model']._get(self.res_model).id,
+            'summary': self.summary,
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_activity.py:519]: DO not hide changes in a move ... Why this change ? Why should we keep summary from a given activity to the next one ? Chances are high this will be unrelated to the new activity.

File: addons/mail/models/mail_activity.py (hunk 1 above)
Comment (tde-banana-odoo) [inline addons/mail/models/mail_activity.py:507]: Could be great to add some tests in ``test_mail`` addons. We notably have some custom test models, you could add an override on ``mail.test.activity`` model and check it is effectively called and taken into consideration.

**Review by tde-banana-odoo — COMMENTED** [review]
First validation !

About commit itself: could you update header and body according ot guidelines ? Notably the "if concatenated with ..." part ?

Many thanks,

File: addons/mail/models/mail_activity.py (hunk 2 above)
Comment (h4818) [reply addons/mail/models/mail_activity.py:519]: It was not my intention to "hide" the change.
Indeed you are right, there is a good chance the summary should be a new one, depending on the next activity - however, the `_onchange_activity_type_id()` will set the summary according to the activity type - and it is called after this line, will therefore overwrite the `summary` value.
This will ONLY apply in the event that there is no summary defined in the activity type.

The reason for this is so that we can have a document request that is named appropriately. Without this change, we have the new document request that is named according to the previous document uploaded. This way, it is named according to the previous document request.

Would you prefer a separate commit for this?

Comment (h4818) [reply addons/mail/models/mail_activity.py:519]: Or alternatively, it could be done in the documents override - and ensure it is only done for document related activities

Comment (h4818) [reply addons/mail/models/mail_activity.py]: removed - I manage this in the documents override instead

**Review by tde-banana-odoo — COMMENTED** [review]
Business code LGTM, missing some tests in test_mail as discussed on discord as it seems action_done is not completely covered by current tests.

File: addons/test_mail/tests/test_mail_activity.py [hunk 3]
```diff
@@ -266,6 +266,27 @@ def test_action_feedback_attachment(self):
             self.assertEqual(attachment.res_id, activity_message.id)
             self.assertEqual(attachment.res_model, activity_message._name)
 
+    def test_chained_activities(self):
+        partner_model = self.env.ref('base.model_res_partner')
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_activity.py:270]: Could be ``Partner = self.env['ir.model']._get('res.partner')`` . That is a more common way to access models.

File: addons/test_mail/tests/test_mail_activity.py [hunk 4]
```diff
@@ -266,6 +266,27 @@ def test_action_feedback_attachment(self):
...
+            'summary': 'Test',
+            'activity_type_id': self.env.ref('test_mail.mail_act_test_chained_1').id,
+            'res_model_id': partner_model.id,
+            'res_id': partner.id,
+        })
+        chained_activity.action_feedback(feedback='Done')
+
+        self.assertFalse(chained_activity.exists())
+
+        next_activity = self.env['mail.activity'].search([('res_model_id','=', partner_model.id), ('res_id','=', partner.id)])
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_activity.py:284]: We could also simply take ``next_activity = partner.activity_ids``, isn't it ? That way we use the interface field.

**Review by tde-banana-odoo — COMMENTED** [review]
Review: LGTM, some small remarks on tests

Note: I think we talked about squashing commits, probably a good idea to do so :) . Could you also update in commit message, use a "Task-XYZ" naming ? That way git does not try to convert the task ID into a commit hash.

Thanks !

File: addons/test_mail/tests/test_mail_activity.py (hunk 4 above)
Comment (h4818) [reply addons/test_mail/tests/test_mail_activity.py]: ok done

File: addons/test_mail/tests/test_mail_activity.py (hunk 3 above)
Comment (h4818) [reply addons/test_mail/tests/test_mail_activity.py]: ok done

**Comment by tde-banana-odoo** [comment]
@robodoo r+
